crates/webidl/src/idl_type.rs

```diff
@@ -88,6 +88,174 @@ pub(crate) enum IdlType<'a> {
     UnknownInterface(&'a str),
 }
 
+/// Indication that the [syn::Type] returned along with this does not tell the whole truth encoded
+/// in the IDL. All its properties are shaped by which information it needs in its `.tell()` method
+/// that converts the gathered information into something that can be put into the docs.
+#[derive(Debug)]
+pub struct Confession {
+    kind: ConfessionKind,
+    actual_type: Option<String>,
+    is_behind_result: bool,
+    has_more_data: bool,
+}
+
+#[derive(Debug)]
+enum ConfessionKind {
+    Promise,
+    Iterable,
+    Union,
+    Callback,
+}
+
+/// Render a syn::Type in such a way that it can be included in rustdoc text.
+///
+/// This generally looks similar to quoting it, but will make a best-effort attempt to create
+/// usable rustdoc links from it or its parts. It also attempts some clean-up like removing any
+/// lavishly used whitespace.
+///
+/// The result is not set in the teletype font common for code examples; it is recommended that the
+/// consumer of this string put it in `<code>...</code>` together with any surrounding text. (If
+/// backticks were used, those would wind up in separate code tags, creating artifacts inbetween
+/// the production boundaries, and moreover the backticks might interfere in unexpected ways).
+///
+/// A more advanced version of this would look into the details of the syn and render them
+/// piecemeal and recursively, to produce things like `[Option]<[::some::Foo]>` (which is unlike
+/// leaving all to rustdoc which would accept a `[Option<::some::Foo>]` but merely render it as
+/// "Option"); for now this captures the easy cases which make up most of webidl.
+fn doc_prettyprint(t: &syn::Type) -> String {
+    use quote::quote;
+
+    fn is_simple(s: &str) -> bool {
+        s
+            // Don't accept any whitespace -- `mut Foo` is not simple
+            .replace(" :: ", "")
+            .replace(":: ", "")
+            .chars().all(|c| c.is_alphanumeric() || c == '_')
+    }
+
+    let quoted = format!("{}", quote! { #t });
```

Comment:
I think it would be better to work over the structure of the `syn::Type` here rather than looking at the stringification of it. That feels sort of brittle and error-prone where we shouldn't really be dealing with too too fancy types here?

Reply:
I've tried going that path -- but it isn't sufficient to superficially look at them, and in the end would require reimplementing most of quote (even to get to the "simple" cases). I had hoped to match on a few versions, but even just looking at Path means that for each path entry it needs to be understood whether that is just a literal or has some arguments that'd need special processing.

Reply:
Hm I'd be surprised if we're getting all sorts of flavorful types here. What sort of `syn::Type` shapes are getting processed here?